ruby-gem/lib/calabash-android/drag_helpers.rb

```diff
@@ -0,0 +1,109 @@
+require 'calabash-android/env'
+
+module Calabash
+  module Android
+    module DragHelpers
+
+      MAX_RETRIES = 7
+      @@monkey_port = nil
+
+      def start_monkey
+        kill_existing_monkey_processes
+        retries = 0
+        loop do
+          port = rand((1025..65535))
+          output = ''
+          fork do
+            output = `#{Env.adb_path} shell monkey --port #{port}`
+          end
+          sleep(2.0)
+          unless output.include? 'Error binding to network socket.'
+            #The shim will always exit, but on success it will not output the line above
+            pids = existing_monkey_pids
+            if pids.count == 1 then
+              return port
+            end
+          end
+          break if (retries += 1) > MAX_RETRIES
+        end
+      end
+
+      def existing_monkey_pids
+        procs = `#{Env.adb_path} shell ps`
+        procs.scan(/.+?\s(?<pid>[0-9]+).+?com.android.commands.monkey\r?\n?/)
+      end
+
+      def kill_existing_monkey_processes
+        existing_monkey_pids.each do |pid|
+          if pid.is_a? Array
+            pid = pid.first
+          end
+          `#{Env.adb_path} shell kill -9 #{pid}`
+        end
+      end
+
+      def verify_connection
+        #lazy instantiation of monkey, since it takes a while
+        @@monkey_port ||= start_monkey
+      end
+
+      def monkey_touch(touch_type, x, y)
+        verify_connection
+        case touch_type
+          when :move
+            perform_action('send_tcp', @@monkey_port, "touch move #{x} #{y}", true)
+          when :down
+            perform_action('send_tcp', @@monkey_port, "touch down #{x} #{y}", true)
+          when :up
+            perform_action('send_tcp', @@monkey_port, "touch up #{x} #{y}", true)
+          else
+            raise "touch_type #{touch_type} is invalid"
+        end
+      end
+
+      def drag_coordinates(from_x, from_y, to_x, to_y, options={})
+        puts "Dragging from #{from_x},#{from_y} to #{to_x},#{to_y}"
+
+        hold_time = options[:hold_time] || 1
+        hang_time = options[:hang_time] || 1
+        steps     = options[:steps] || 10
+
+        monkey_touch(:down, from_x, from_y)
+        sleep(hold_time)
+
+        x_delta = to_x - from_x
+        y_delta = to_y - from_y
+
+        (1..steps).each do |i|
+          sleep(0.2)
```

Comment:
Sorta - empirically, it didn't seem to work without some sleep. But it's also possibly my code was just buggy at the time - I'll revisit.